alacritty_terminal/src/ansi.rs

```diff
@@ -25,60 +25,88 @@ use vte;
 use crate::term::color::Rgb;
 
 // Parse color arguments
-//
-// Expect that color argument looks like "rgb:xx/xx/xx" or "#xxxxxx"
-fn parse_rgb_color(color: &[u8]) -> Option<Rgb> {
-    let mut iter = color.iter();
 
-    macro_rules! next {
-        () => {
-            iter.next().map(|v| *v as char)
-        };
+// Parse a color argument in the form "#xxxxxx"
+// The use of this format is no longer encouraged by xparsecolor
+fn parse_legacy_rgb_color(color: &[u8]) -> Option<Rgb> {
+    if color.len() < 7 || color[0] != b'#' {
+        return None;
     }
 
+    let mut iter = color.iter().map(|v| (*v as char));
+    iter.next();
+
     macro_rules! parse_hex {
         () => {{
-            let mut digit: u8 = 0;
-            let next = next!().and_then(|v| v.to_digit(16));
-            if let Some(value) = next {
-                digit = value as u8;
-            }
+            let mut hex: u8 = match iter.next().and_then(|v| v.to_digit(16)) {
+                Some(digit) => digit as u8,
+                _ => return None,
+            };
+            hex <<= 4;
+            hex += match iter.next().and_then(|v| v.to_digit(16)) {
+                Some(digit) => digit as u8,
+                _ => return None,
+            };
+            hex
+        }};
+    }
+
+    let r = parse_hex!();
+    let g = parse_hex!();
+    let b = parse_hex!();
 
-            let next = next!().and_then(|v| v.to_digit(16));
-            if let Some(value) = next {
-                digit <<= 4;
-                digit += value as u8;
+    Some(Rgb { r, g, b })
+}
+
+fn is_hex_separator(ch: Option<char>) -> bool {
+    ch == None || ch == Some('/') || ch == Some('\x07')
+}
+
+// Expect that color argument looks like "rgb:x[xxx]/x[xxx]/x[xxx]" or "#xxxxxx"
+fn parse_rgb_color(color: &[u8]) -> Option<Rgb> {
+    let mut iter = color.iter().map(|v| *v as char);
+
+    macro_rules! parse_hex {
+        () => {{
+            let mut hex: u8 = match iter.next().and_then(|v| v.to_digit(16)) {
+                Some(digit) => digit as u8,
+                _ => return None,
+            };
+            hex <<= 4;
+
+            let next = iter.next();
+            if !is_hex_separator(next) {
+                if let Some(digit) = next.and_then(|v| v.to_digit(16)) {
+                    hex += digit as u8;
+                }
+                if (0..3).find(|_| is_hex_separator(iter.next())) == None {
+                    return None;
+                }
             }
-            digit
+
+            hex
         }};
```

Comment:
Since this is pretty much a direct copy of the legacy rgb code, we should extract it to a separate function.

The `/` separation should have no impact on the legacy color parsing anyways, so we can just have one function that handles both.